src/electronic-id.cpp

```diff
@@ -112,7 +107,46 @@ const std::map<byte_vector, ElectronicIDConstructor> SUPPORTED_ATRS {
      constructor<ElectronicID::Type::CzeEID>},
 };
 
-inline std::string byteVectorToHexString(const byte_vector& bytes)
+// Holds ATR pattern, mask, and constructor for variable ATR cards.
+struct MaskedATREntry
+{
+    // Single template parameter enforces equal size pattern and mask arrays at compile time.
+    template <size_t N>
+    constexpr MaskedATREntry(const byte_type (&_pat)[N], const byte_type (&_mask)[N],
+                             ElectronicIDConstructor&& _constructor) :
+        pattern(std::begin(_pat), std::end(_pat)), mask(std::begin(_mask), std::end(_mask)),
+        constructor(std::move(_constructor))
+    {
+    }
+
+    bool operator==(const byte_vector& atr) const
+    {
+        if (atr.size() != pattern.size()) {
+            return false;
+        }
+
+        for (size_t i = 0; i < atr.size(); ++i) {
+            if ((atr[i] & mask[i]) != (pattern[i] & mask[i])) {
+                return false;
+            }
+        }
+
+        return true;
```

Comment:
```suggestion
        return std::equal(atr.cbegin(), atr.cend(), pattern.cbegin(), pattern.cend(),
                          [mask_ptr = mask.data()](byte_type a, byte_type p) mutable {
                              bool result = (a & *mask_ptr) == (p & *mask_ptr);
                              ++mask_ptr;
                              return result;
                          });
```

Reply:
Although std algorithms are generally preferable, in this case the for loops looks more clear and straightforward in my opinion, but let's go with this.